azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/AzureAccount.java

```diff
@@ -5,38 +5,135 @@
+
+    public List<AccountEntity> getAvailableAccounts() {
+        Map<AuthType, IAccountEntityBuilder> allBuilders = buildProfilerBuilders(environment);
+        return allBuilders.values().stream().map(builder -> builder.build()).collect(Collectors.toList());
     }
 
-    /**
-     * @return account if authenticated
-     * @throws AzureToolkitAuthenticationException if failed to authenticate
-     */
-    public Account login(AccountEntity entity) throws AzureToolkitAuthenticationException {
-        final Account account = new Account(); // TODO: login and create Account instance;
-        this.account = account;
-        return this.account;
+    public void login(AuthConfiguration auth) {
+        environment = auth.getEnvironment();
+        boolean forceSPLogin = auth.getType() == AuthType.SERVICE_PRINCIPAL;
+        if (forceSPLogin || auth.getType() == AuthType.AUTO) {
+            boolean isSPConfigurationPresent = !StringUtils.isAllBlank(auth.getCertificate(), auth.getKey(),
+                    auth.getCertificatePassword());
+            try {
+                ValidationUtil.validateAuthConfiguration(auth);
+            } catch (InvalidConfigurationException e) {
+                if (forceSPLogin) {
+                    Lombok.sneakyThrow(e);
+                    return;
+                }
+                if (isSPConfigurationPresent) {
+                    logger.warning("Cannot login through 'SERVICE_PRINCIPAL' due to invalid configuration: " + e.getMessage());
+                }
+            }
+
+            ServicePrincipalAccountEntityBuilder builder = new ServicePrincipalAccountEntityBuilder(auth);
+            AccountEntity entity = builder.build();
+            if (entity.isAuthenticated()) {
+                login(entity);
+                return;
+            }
+        }
+        Map<AuthType, IAccountEntityBuilder> allBuilders = buildProfilerBuilders(environment);
+        if (allBuilders.containsKey(auth.getType())) {
+            loginByBuilder(allBuilders.get(auth.getType()));
+            return;
+        }
+
+        for (IAccountEntityBuilder builder : allBuilders.values()) {
+            if (loginByBuilder(builder)) {
+                return;
+            }
+        }
+    }
+
+    public void login(AccountEntity accountEntity) throws AzureToolkitAuthenticationException {
+        Objects.requireNonNull(accountEntity, "accountEntity is required.");
+        this.account = new Account();
+        final AccountEntity entity = new AccountEntity();
+        entity.setMethod(accountEntity.getMethod());
+        entity.setAuthenticated(accountEntity.isAuthenticated());
+        entity.setEnvironment(accountEntity.getEnvironment());
+        entity.setEmail(accountEntity.getEmail());
+        entity.setTenantIds(accountEntity.getTenantIds());
+        this.account.setEntity(entity);
+        this.account.setCredentialBuilder(accountEntity.getCredentialBuilder());
+        if (!accountEntity.isAuthenticated()) {
```

Comment:
Check whether account entity is authenticated first?

Reply:
Login is likely to change the state(Auth method) of AzureAccount, if login fails, the account is left with the failure login with
authenticated = false